[NETBEANS-2172] & [NETBEANS-2398] Fix for web.xml errors

peedeeboy opened the pull request:
*What this pull request does*
Fixes the errors when creating a web.xml file.  Before this fix, the parser will warn about the <web-app> tag not being recognized and XML validation will fail due to the .xsd files not being known

*Root Cause Analysis*
It seems that the JavaEE Deployment Descriptor .xsd files used to be a part of the JavaEE Deployment Descriptor loader module, but are now part of the J2EE Deployment Descriptor API module.

Possibly they moved / removed & re-added during the licencing of JavaEE for Apache?  I'm assuming that their new location is the correct/intended one.

The Netbeans XML catalog "Enterprise Deployment Descriptor Catalog" was still pointing to the .xsd files being part of the JEE DD Loader module, so various things were falling over when trying to access them

* Files Changed*
* /enterprise/j2ee.ddloaders/src/org/netbeans/modules/j2ee/ddloaders/catalog/EnterpriseCatalog.java
Resource Path constant changed from nbres:/org/netbeans/modules/j2ee/ddloaders/catalog/resources/ to nbres:/org/netbeans/modules/j2ee/dd/impl/resources/

*Testing*
* Make sure Java EE plugin activated
* Tools-> DTD & XML Schemas -> Expand Enterprise Deployment Descriptors catalog
* Chose any .xsd and click Open in Editor
Before this PR: An error will be thrown "Could not connect to URL nbres:/org/netbeans/modules/j2ee/ddloaders/catalog/resources/web-app_3_1.xsd. No such resource was found."
After this PR: .xsd file will be opened in Netbeans Editor
* Create a new Java Web Project, add a new web.xml file
Before this PR: File will be highlighted with errors / validate XML will fail
After this PR: No errors

geertjanw commented:
Excellent, thanks so much for investigating and fixing this.

matthiasblaesing commented:
Thank you - the fix looks ok. I agree with the assessment and the fix to use the dtds/xsd from `j2ee.dd`, as `j2ee.ddloaders` already depends on it. However, the collection is not complete. `EnterpriseCatalog` references theses files:

- j2ee_web_services_1_1.xsd
- javaee_web_services_1_2.xsd
- javaee_web_services_1_3.xsd
- javaee_web_services_1_4.xsd
- connector_1_5.xsd
- connector_1_7.xsd
- web-app_2_2.dtd
- web-app_2_3.dtd
- beans_1_0.xsd
- beans_1_1.xsd
- batchXML_1_0.xsd
- jobXML_1_0.xsd

which are not present in the new location. I suggest to fetch the missing file from their origin: https://www.oracle.com/webfolder/technetwork/jsc/xml/ns/javaee/index.html and add them to the directory. If I remember correctly all schema files published by Oracle where either GPLv2-CPE/CDDL1.0 or GPLv2-CPE/CDDL1.1, so as with the other schemas we elect the CDDL variant, add entries for the files to the licenseinfo.xml of the corresponding module and it should be done. It would be great if you could update the PR with that change.

peedeeboy commented:
Thanks for the feedback @matthiasblaesing !  Looks like none of the existing .xsd files in the `j2ee.dd` module are referenced in the `licenceinfo.xml` file either.  I'll add in the proper entries for those as I go through.

Away for a few days.  Will have this ready early next week :)

peedeeboy commented:
Ok, couple of quick questions:

1) jobXML_1_0.xsd makes no mention of licence.  What to put in licenseinfo.xml for this file?

2) I've added a new <fileset> section to licenseinfo.xml for CDDL-1.0 licensed files, however all the existing .xsd files were in a <fileset> section for CDDL1.1, even if their license header suggest CDDL-1.0?  Should I do likewise, or keep the separate CDDL1.0 section?

matthiasblaesing commented:
For 1: jobXML_1_0.xsd is bad - I suggest to skip that file and remove the entry from the schema catalog (EnterpiseCatalog.java).
For 2:  File entries, that are not correctly categorized, should be corrected. If you see CDDL-1.0 files declared to be CDDL-1.1 please update them (yes you'll need a separate entry for them).

peedeeboy commented:
- Removed `jobXML_1_0.xsd` file and the reference to it in `EnterpriseCatalog.java`
- Added missing .xsd / .mdd files already referenced in `EnterpriseCatalog.java`
- Updated `licenseinfo.xml` to include missing .xsd / .mdd files & moved existing CDDL-1.0 license files to new section

Commits squashed into one.

Hope that helps :)

peedeeboy commented on enterprise/j2ee.dd/licenseinfo.xml:
```diff
@@ -24,6 +24,9 @@
+        <file>src/org/netbeans/modules/j2ee/dd/impl/resources/batchXML_1_0.xsd</file>
+        <file>src/org/netbeans/modules/j2ee/dd/impl/resources/beans_1_0.xsd</file>
+        <file>src/org/netbeans/modules/j2ee/dd/impl/resources/beans_1_1.xsd</file>
         <license ref="Apache-2.0-ASF" />
```
No worries :)  It's a learning experience for me, and good information to know for the future.

- Have separated those three files into a new `<fileset>` section for `Apache-2.0 license`

matthiasblaesing commented:
Thank you for your work. I merged this as:
https://github.com/apache/netbeans/commit/d172c20fa5ed9e3aad4a025879f0a94763fb4905
I removed the entries for the Apache-2.0 licensed files. I had overlooked, that the rat report, that checks the licenses recoginsed the files correctly and Apache-2.0 is Category A and thus is no problem for the rat check.
